external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSinkSuite.scala

```diff
@@ -0,0 +1,413 @@
+        CheckAnswer((1, 1), (2, 2), (3, 3), (1, 2), (2, 3), (3, 4)),
+        StopStream
+      )
+      writer.stop()
+    }
+  }
+
+
+  test("write data with bad schema") {
+    withTempDir { checkpointDir =>
+      val input = MemoryStream[String]
+      val topic = newTopic()
+      testUtils.createTopic(topic)
+
+      /* No topic field or topic option */
+      var writer: StreamingQuery = null
+      var ex = intercept[StreamingQueryException] {
+        writer = input.toDF()
+          .selectExpr("value as key", "value")
+          .writeStream
+          .format("kafka")
+          .option("checkpointLocation", checkpointDir.getCanonicalPath)
+          .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+          .queryName("kafkaNoTopicFieldStream")
+          .start()
+        input.addData("1", "2", "3", "4", "5")
+        writer.processAllAvailable()
+      }
+      writer.stop()
+      assert(ex.getMessage
+        .toLowerCase
+        .contains("topic option required when no 'topic' attribute is present"))
+
+      /* No value field */
+      ex = intercept[StreamingQueryException] {
+        writer = input.toDF()
+          .selectExpr(s"'$topic' as topic", "value as key")
+          .writeStream
+          .format("kafka")
+          .option("checkpointLocation", checkpointDir.getCanonicalPath)
+          .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+          .queryName("kafkaNoValueFieldStream")
+          .start()
+        input.addData("1", "2", "3", "4", "5")
+        writer.processAllAvailable()
+      }
+      writer.stop()
+      assert(ex.getMessage.toLowerCase.contains("required attribute 'value' not found"))
+    }
+  }
+
+  test("write data with valid schema but wrong types") {
+    withTempDir { checkpointDir =>
+      val input = MemoryStream[String]
+      val topic = newTopic()
+      testUtils.createTopic(topic)
+
+      /* value field wrong type */
+      var writer: StreamingQuery = null
+      var ex = intercept[StreamingQueryException] {
+        writer = input.toDF()
+          .selectExpr(s"'$topic' as topic", "CAST(value as INT) as value")
+          .writeStream
+          .format("kafka")
+          .option("checkpointLocation", checkpointDir.getCanonicalPath)
+          .option("kafka.bootstrap.servers", testUtils.brokerAddress)
+          .queryName("kafkaIntValueFieldStream")
+          .start()
+        input.addData("1", "2", "3", "4", "5")
+        writer.processAllAvailable()
+      }
+      writer.stop()
+      assert(ex.getMessage.toLowerCase.contains(
+        "value attribute type must be a string or binarytype"))
+
+      /* key field wrong type */
+      ex = intercept[StreamingQueryException] {
+        writer = input.toDF()
+          .selectExpr(s"'$topic' as topic", "CAST(value as INT) as key", "value")
+          .writeStream
+          .format("kafka")
+          .option("checkpointLocation", checkpointDir.getCanonicalPath)
```

Comment:
why do you need checkpoint location for this test? simplify each test as much as possible with as little requirements as possible. 

and a lot of this code duplicate. make internal functions to reduce duplication.